packages/utils/node-resolver-rs/src/cache.rs

```diff
@@ -0,0 +1,169 @@
+  ops::Deref,
+  path::{Path, PathBuf},
+};
+
+use elsa::FrozenMap;
+use typed_arena::Arena;
+
+use crate::{
+  fs::{FileSystem, OsFileSystem},
+  package_json::{PackageJson, SourceField},
+  tsconfig::{TsConfig, TsConfigWrapper},
+  ResolverError,
+};
+
+#[derive(Default)]
+pub struct Cache<Fs = OsFileSystem> {
+  pub fs: Fs,
+  // This stores file content strings, which are borrowed when parsing package.json and tsconfig.json files.
+  arena: Arena<Box<str>>,
+  // These map paths to parsed config files. They aren't really 'static, but Rust doens't have a good
+  // way to associate a lifetime with owned data stored in the same struct. We only vend temporary references
+  // from our public methods so this is ok for now. FrozenMap is an append only map, which doesn't require &mut
+  // to insert into. Since each value is in a Box, it won't move and therefore references are stable.
+  packages: FrozenMap<PathBuf, Box<Result<PackageJson<'static>, ResolverError>>>,
+  tsconfigs: FrozenMap<PathBuf, Box<Result<TsConfigWrapper<'static>, ResolverError>>>,
+}
+
+// Special Cow implementation for a Cache that doesn't require Clone.
+pub enum CacheCow<'a, Fs> {
+  Borrowed(&'a Cache<Fs>),
+  Owned(Cache<Fs>),
+}
+
+impl<'a, Fs> Deref for CacheCow<'a, Fs> {
+  type Target = Cache<Fs>;
+
+  fn deref(&self) -> &Self::Target {
+    match self {
+      CacheCow::Borrowed(c) => *c,
+      CacheCow::Owned(c) => c,
+    }
+  }
+}
+
+#[derive(Debug, Clone, PartialEq, serde::Serialize)]
+pub struct JsonError {
+  pub path: PathBuf,
+  pub line: usize,
+  pub column: usize,
+  pub message: String,
+}
+
+impl JsonError {
+  fn new(path: PathBuf, err: serde_json::Error) -> JsonError {
+    JsonError {
+      path,
+      line: err.line(),
+      column: err.column(),
+      message: err.to_string(),
+    }
+  }
+}
+
+impl<Fs: FileSystem> Cache<Fs> {
+  pub fn new(fs: Fs) -> Self {
+    Self {
+      fs,
+      arena: Arena::new(),
+      packages: FrozenMap::new(),
+      tsconfigs: FrozenMap::new(),
+    }
+  }
+
+  pub fn read_package<'a>(&'a self, path: Cow<Path>) -> Result<&'a PackageJson<'a>, ResolverError> {
+    if let Some(pkg) = self.packages.get(path.as_ref()) {
+      return clone_result(pkg);
+    }
+
+    fn read_package<Fs: FileSystem>(
+      fs: &Fs,
+      arena: &Arena<Box<str>>,
+      path: PathBuf,
+    ) -> Result<PackageJson<'static>, ResolverError> {
+      let data = read(fs, arena, &path)?;
+      let mut pkg = PackageJson::parse(path.clone(), data).map_err(|e| JsonError::new(path, e))?;
+
+      // If the package has a `source` field, make sure
+      // - the package is behind symlinks
+      // - and the realpath to the packages does not includes `node_modules`.
+      // Since such package is likely a pre-compiled module
+      // installed with package managers, rather than including a source code.
+      if !matches!(pkg.source, SourceField::None) {
+        let realpath = fs.canonicalize(&pkg.path)?;
+        if realpath == pkg.path
```

Comment:
Not sure if maybe we should change this logic for pnpm... #5784

Reply:
I think this should work because the realpath will indeed contain node_modules: `/Users/niklas/Desktop/test/node_modules/.pnpm/file+sub/node_modules/sub/index.js`? Or which case are you thinking of that doesn't work with this logic?

Reply:
Ah ok I wasn't sure if node_modules was still in the path